File: Yandex/Practicum/Algorightms_and_Data_Structures/7_sprint/E._Алла_на_Алгосах/solution.cpp
#include <iostream>
#include <vector>


void solution() {
    /*
    Time Complexity O(NM)
    Memory Complexity O(M)
    */
    int x;
    std::scanf("%d", &x);
    int k;
    std::scanf("%d", &k);
    std::vector<int> coins (k);
    for (int i = 0; i < k; ++i) std::cin >> coins[i];
    std::vector<int> dp (x + 1, 10001);
    dp[0] = 0;
    for (int exchange = 1; exchange <= x; ++exchange) {
        for (int& coin : coins) {
            if (coin > exchange) continue;
            dp[exchange] = std::min(dp[exchange], dp[exchange - coin] + 1);
        }
    }
    std::cout << (dp[x] != 10001? dp[x] : -1) << "\n";
}


int main() {
    solution();
}

File: Yandex/Practicum/Algorightms_and_Data_Structures/7_sprint/E._Алла_на_Алгосах/solution.py
import sys


def solution() -> None:
    """
    Time Complexity O(NM)
    Memory Complexity O(M)
    """
    x: int = int(sys.stdin.readline().rstrip())
    k: int = int(sys.stdin.readline().rstrip())
    coins: list[int] = list(map(int, sys.stdin.readline().rstrip().split()))
    dp: list[int] = [10_001] * (x + 1)
    dp[0] = 0
    for exchange in range(1, x + 1):
        for coin in coins:
            if coin > exchange: continue
            dp[exchange] = min(dp[exchange], dp[exchange - coin] + 1)
    sys.stdout.write(f'{dp[x] if dp[x] != 10_001 else -1}\n')


if __name__ == '__main__':
    solution()